### PR TITLE
fix tensorflow test hanging issue

### PR DESCRIPTION
**Pull request by frank-dong-ms**
fixed 2 issue of tensorflow tests:
1. fix issue in Resource Manager, use ReadAsync instead of Read to let thread timeout corrently
2. move download ImageSet to constructor of test class to avoid duplicate download for different tests

### CODE REVIEW

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs [hunk 1]
```diff
@@ -1250,25 +1263,17 @@ public void TensorFlowStringTest()
...
-                "images");
-
-            //Download the image set and unzip
-            string finalImagesFolderName = DownloadImageSet(
-                imagesDownloadFolderPath);
-
-            string fullImagesetFolderPath = Path.Combine(
-                imagesDownloadFolderPath, finalImagesFolderName);
+            // set timeout to 3 minutes, download sometimes will stuck so set smaller timeout to retry download
+            string timoOutOldValue = Environment.GetEnvironmentVariable(ResourceManagerUtils.TimeoutEnvVariable);
```
Comment (harishsk) [inline test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs:1269]: Why do we want do this only for these tests and not for all tests? We can choose a smaller timeout and higher retry count for all downloads.

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs [hunk 2]
```diff
@@ -1338,25 +1346,17 @@ internal bool ShouldReuse(string workspacePath, string trainSetBottleneckCachedV
...
+            Environment.SetEnvironmentVariable(ResourceManagerUtils.TimeoutEnvVariable, (3 * 60 * 1000).ToString());
```
Comment (harishsk) [inline test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs:1354]: Will setting the environment variables here and resetting it at the end of the test have any impact?
It appears that you have moved the download code to the constructor. But the timeout is not being set there. So by the time the code gets here wouldn't the download already have completed with a different timeout value?

File: src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs [hunk 3]
```diff
@@ -127,14 +127,22 @@ public async Task<ResourceDownloadResults> EnsureResourceAsync(IHostEnvironment
...
+                        downloadResult += thisDownloadResult + @"\n";
 
-                await Task.Delay(10 * 1000);
+                    await Task.Delay(10 * 1000);
+                }
+                catch (Exception ex)
+                {
+                    // ignore any Exception and retrying download
+                    ch.Warning($"{i+1} - th try: Dowload {fileName} from {url} fail with exception {ex.Message}");
+                }
```
Comment (harishsk) [inline src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs:145]: We have lots of trouble with failed downloads. Can we make the exceptions more visible? 

That is, we should log all the failures as you have done. But if in the end the download still fails after the retries, how about throwing an exception that download failed even after retries?

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs (hunk 1 above)
Comment (frank-dong-ms) [reply test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs:1269]: this setting (timeout) is hardcoded in our source code as 10 minutes but not a option that user can set at https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Vision/ImageClassificationTrainer.cs#L1321

this setting can be override by set the env variable I'm doing this as if everything is fine 3 minutes is enough if something go wrong then we hope this to fail fast as we are retrying 5 times.

we are not seeing other tests have similar issue so I only do this in affected tests.

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs (hunk 2 above)
Comment (frank-dong-ms) [reply test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs:1354]: They are downloading different things.

In constructor we are downloading some image set and all tests use same images so do need to download duplicate images and I move this to constructor. this is refine the code performance and remove unnecessary downloads.

this env variable is to set timeout for some metadata for test and metadata download is causing the hanging.

File: src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs (hunk 3 above)
Comment (frank-dong-ms) [reply src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs:145]: yes, we indeed throw exception if retry still fails at line: https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.TensorFlow/TensorflowUtils.cs#L190

File: src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs [hunk 4]
```diff
@@ -268,21 +278,28 @@ private Exception DownloadResource(IHostEnvironment env, IChannel ch, WebClient
                         size = 10000000;
 
                     long printFreq = (long)(size / 10.0);
-                    var buffer = new byte[4096];
+                    var buffer = new byte[blockSize];
                     long total = 0;
-                    int count;
+
+                    var task = s.ReadAsync(buffer, 0, blockSize);
+                    task.Wait(ct);
```
Comment (harishsk) [inline src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs:285]: ReadAsync allows you to pass a CancellationToken directly. You can choose to use that.

**Pull request merged**